File: haskell/src/Solutions/Euler081.hs
module Solutions.Euler081 where
import Utils.Input(getInput)
import Control.Monad
import Control.Monad.ST
import Data.Array
import Data.Array.ST

type Matrix = Array (Int, Int) Int

ixs :: Int -> [(Int, Int)] -- this is intentionally transposed
ixs n = [(y, x) | x <- [0..n - 1], y <- [0..n - 1]]

parseRow :: String -> [Int]
parseRow s = read $ "[" ++ s ++ "]"

parseMatrix :: String -> Matrix
parseMatrix s =
    let rows = map parseRow $ lines s
        n = length rows
        associates = zip (ixs n) (concat rows)
    in array ((0, 0), (n - 1, n - 1)) associates

findMinPath :: Matrix -> Int
findMinPath matrix = let n = snd $ snd $ bounds matrix in runST $ do
    arr <- newArray ((0, 0), (n, n)) 0 :: ST s (STArray s (Int, Int) Int)
    writeArray arr (0, 0) $ matrix ! (0, 0)
    forM_ [1..n] $ \x -> do
        readArray arr (x - 1, 0) >>= writeArray arr (x, 0) . (+(matrix ! (x, 0)))
        readArray arr (0, x - 1) >>= writeArray arr (0, x) . (+(matrix ! (0, x)))
    forM_ [1..n] $ \x -> forM_ [1..n] $ \y -> do
        v1 <- readArray arr (x - 1, y)
        v2 <- readArray arr (x, y - 1)
        writeArray arr (x, y) $ matrix ! (x, y) + min v1 v2
    readArray arr (n, n)

euler081 :: IO String
euler081 = show . findMinPath . parseMatrix <$> getInput 81

File: rust/src/solutions/euler081.rs
use crate::utils::input::get_input;

pub fn parse_matrix(s: &str) -> Vec<Vec<usize>> {
    s.lines()
        .map(|line| {
            line.split(',')
                .map(|c| c.parse::<usize>().unwrap())
                .collect()
        })
        .collect()
}

pub fn euler081() -> String {
    let mut arr = parse_matrix(&get_input(81));
    let l = arr.len();
    for i in 1..l {
        arr[i][0] += arr[i - 1][0];
        arr[0][i] += arr[0][i - 1];
    }
    for i in 1..l {
        for j in 1..l {
            arr[i][j] += std::cmp::min(arr[i - 1][j], arr[i][j - 1]);
        }
    }
    arr[l - 1][l - 1].to_string()
}
